src/ecephys_etl/modules/vbn_create_stimulus_table/create_stim_table.py

```diff
@@ -468,12 +527,392 @@ def check_behavior_and_replay_pkl_match(
+                    (pd_interval_duration / expected_vsync_duration)
+                )
+            ) - num_vsyncs_per_diode_toggle
+        )
+
+        # If our total_delay is more than we would expect from just long vsyncs
+        # then extra frame to monitor delay occurred
+        extra_frame_to_monitor_delay = 0
+
+        if total_delay > num_long_vsyncs:
+            print(
+                """Extra delay between frame time
+                 and monitor display time detected"""
+            )
+
+            # Delay attributed to hardware/software factors that delay time
+            # to monitor display (in units of 'vsyncs') must then be:
+            extra_frame_to_monitor_delay = total_delay - num_long_vsyncs
+
+        # Number of actual frames/vsyncs that would fit
+        # in a photodiode switch interval
+        local_expected_vsync_duration = (
+            pd_interval_duration / (num_vsyncs_per_diode_toggle + total_delay)
+        )
+
+        if total_delay > 0:
+            # Correct for variability in vsync times
+            variance_reduced_frame_diffs = (
+                np.round(
+                    np.diff(relevant_vsyncs) / local_expected_vsync_duration
+                )
+            )
+
+            # NJM - Want first vsync to happen at diode transition
+            # NJM - Is the 0th vsync happening before or after first
+            # photodiode transition? There could be 1-off error (double check)
+            # Will need to check empirically when implementing
+            result = (
+                variance_reduced_frame_diffs * local_expected_vsync_duration
+            )
+
+            corrected_relevant_vsyncs = np.insert(
+                np.cumsum(result),
+                0,
+                0
+            )
+
+            # Then correct for extra_frame_to_monitor_delay if there was any
+            # Assume that if there was a change
+            # in monitor lag, it was after the long frame
+            longest_ind = np.argmax(relevant_vsync_durations) + 1
+            corrected_relevant_vsyncs[longest_ind:] += (
+                extra_frame_to_monitor_delay * local_expected_vsync_duration
+            )
+
+            set_corrected_times(
+                corrected_frame_times,
+                vsync_slice,
+                start_time,
+                corrected_relevant_vsyncs
+            )
+
+        else:
+
+            frame_diffs = np.ones(cycle-1)
+            corrected_relevant_vsyncs = np.insert(
+                np.cumsum(frame_diffs * local_expected_vsync_duration),
+                0,
+                0
+            )
+
+            set_corrected_times(
+                corrected_frame_times,
+                vsync_slice,
+                start_time,
+                corrected_relevant_vsyncs
+            )
+
+    # Now deal with leftover vsyncs that occur after the last diode transition
+    # Just take the global frame duration for these
+    leftover_vsyncs_start_ind = (
```

Comment:
This ends up being too fragile. If for some reason camstim misses the last diode transition (or it's trimmed off), then you can have a situation where you have 60 leftover frames, which breaks this (since the mod will be 0). I've seen this for one session so far. Better to find the leftover frames by taking the total frames in this partition minus the ones you've allocated so far in the loop above.